dali/operators/generic/roi_random_crop.cc

```diff
@@ -0,0 +1,207 @@
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  BatchRNG<std::mt19937> rngs_;
+
+  ArgValue<int, 1> roi_start_;
+  ArgValue<int, 1> roi_end_;
+  ArgValue<int, 1> roi_shape_;
+  ArgValue<int, 1> crop_shape_;
+  ArgValue<int, 1> in_shape_arg_;
+
+  TensorListShape<> in_shape_;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+ROIRandomCropCPU::ROIRandomCropCPU(const OpSpec &spec)
+    : Operator<CPUBackend>(spec),
+      rngs_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<int64_t>("max_batch_size")),
+      roi_start_("roi_start", spec),
+      roi_end_("roi_end", spec),
+      roi_shape_("roi_shape", spec),
+      crop_shape_("crop_shape", spec),
+      in_shape_arg_("in_shape", spec) {
+  if ((roi_end_.IsDefined() + roi_shape_.IsDefined()) != 1)
+    DALI_FAIL("Either ROI end or ROI shape should be defined, but not both");
+}
+
+bool ROIRandomCropCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<CPUBackend> &ws) {
+  int nsamples = spec_.HasTensorArgument("crop_shape") ?
+                     ws.ArgumentInput("crop_shape").size() :
+                     ws.GetRequestedBatchSize(0);
+  crop_shape_.Acquire(spec_, ws, nsamples, true);
+  int ndim = crop_shape_[0].shape[0];
+
+  TensorShape<1> sh{ndim};
+  roi_start_.Acquire(spec_, ws, nsamples, sh);
+  if (roi_end_.IsDefined()) {
+    roi_end_.Acquire(spec_, ws, nsamples, sh);
+  } else {
+    assert(roi_shape_.IsDefined());
+    roi_shape_.Acquire(spec_, ws, nsamples, sh);
+  }
+
+  in_shape_.shapes.clear();
+  if (in_shape_arg_.IsDefined() || ws.NumInput() == 1) {
+    if (in_shape_arg_.IsDefined() && (ws.NumInput() == 1)) {
+      DALI_FAIL("``in_shape`` argument is incompatible with providing an input.")
+    }
+    if (in_shape_arg_.IsDefined()) {
+      in_shape_.resize(nsamples, ndim);
+      in_shape_arg_.Acquire(spec_, ws, nsamples, sh);
+      for (int s = 0; s < nsamples; s++) {
+        auto sample_sh = in_shape_.tensor_shape_span(s);
+        for (int d = 0; d < ndim; d++) {
+          sample_sh[d] = in_shape_arg_[s].data[d];
+        }
+      }
+    } else {
+      auto &in = ws.template InputRef<CPUBackend>(0);
+      in_shape_ = in.shape();
+    }
+
+    for (int s = 0; s < nsamples; s++) {
+      auto sample_sh = in_shape_.tensor_shape_span(s);
+      for (int d = 0; d < ndim; d++) {
+        DALI_ENFORCE(sample_sh[d] >= crop_shape_[s].data[d],
+                     "Cropping shape can't be bigger than the input shape.");
+      }
+      if (roi_shape_.IsDefined()) {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 &&
+                       sample_sh[d] >= (roi_start_[s].data[d] + roi_shape_[s].data[d]),
+                       "ROI can't be out of bounds.");
+        }
+      } else {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 && sample_sh[d] >= roi_end_[s].data[d],
```

Comment:
Do we need to bother with `roi_end_` here. Maybe it should be just `roi_start_` and `roi_shape_`, and values from `roi_end` should be used to infer `roi_shape_` if not provided directly.

Reply:
It's easier to check `end` than `start+shape`?